---
title: "Fisces LULU addition"
author: "Lauren Howe-Kerr"
date: "11/18/19"
output: pdf_document
---
LULU- based partly on an idea we’ve had for a while, which is that OTUs that co-occur are more likely to be intragenomic variants. The nice thing about this package is that it also takes sequence similarity into account when trying to assign potential “errors” (such as intragenomic variants) to their parent sequences. So for something to be assigned as a ‘daughter’ it both has to sufficiently co-occur and exhibit some level of sequence similarity.

Need these two inputs for LULU:

*OTU fasta file input name:* asvsforLULU_fisces_Nov2019.fasta
•	Note: I had to change header names: “>sq1” etc and not “>sq1; size=1321321;”
do this in terminal: cut -d ';' -f 1 seqtab.nochim_FISHPOO_7June2019.fasta > asvsforLULU_fisces_Nov2019.fasta

*DADA output file name:* OutputDADA_AllOTUs_FISHPOO_7June2019.csv

Necessary pre-steps after producing ASV table and associated fasta file:
Produce a match list using BLASTn

#######IN TERMINAL#########
First produce a blastdatabase with the OTUs
>makeblastdb -in asvsforLULU_fisces_Nov2019.fasta -parse_seqids -dbtype nucl

Then blast the OTUs against the database to produce the match list 
>blastn -db asvsforLULU_fisces_Nov2019.fasta -outfmt '6 qseqid sseqid pident' -out match_list_Fisces.txt -qcov_hsp_perc 90 -perc_identity 84 -query asvsforLULU_fisces_Nov2019.fasta

```{r}
library(devtools)
#install_github("tobiasgf/lulu") 
library(lulu)

alldat<-read.csv("/Users/laurenhowe-kerr/Dropbox/FISHPOO/unzipped/AllPoops/OutputDADA_AllOTUs_FISHPOO_7June2019.csv")
head(alldat)

matchList<-read.table("/Users/laurenhowe-kerr/Dropbox/FISHPOO/unzipped/AllPoops/match_list_Fisces.txt")
head(matchList)
```


```{r}
#Reformat ASV table to desired LULU format
rownames(alldat)<-alldat$SampleID
ASVs<-data.frame(t(alldat[,2:749])) #to remove the sequence ID column
head(ASVs)


#Now, run the LULU curation

curated_result <- lulu(ASVs, matchList) #retains 22 OTUs at 97% similarity; still retains 19 OTUs at 84% similarity and clustered OTUs are low abundance variants. Lets go with the more conservative clustering. 
summary(curated_result)

curated_result$minimum_match
curated_result$minimum_relative_cooccurence

###I AM 100000% SURE THERE IS A BETTER WAY TO GET THIS SHIT FORMATTED BUT THIS IS WHAT I FIGURED OUT, LMK HOW TO DO IT BETTER PLZZZZ

#make the results into a dataframe
alldat<-cbind(data.frame(curated_result$curated_table))
head(alldat)

#make into a CSV so that I can add a column name ("Sq.id"") to the "sq" columns for merging
write.csv(alldat,file="~/Dropbox/FISHPOO/unzipped/AllPoops/alldat_fisces_luluNOV19.csv",row.names=TRUE,quote=FALSE)
#now add "Sq.id" to name the first column, saved as "mod" document
alldat.mod <- read.csv("~/Dropbox/FISHPOO/unzipped/AllPoops/alldat_fisces_luluNOV19_mod.csv")

#make OTU fasta file into a dataframe (for merging with dataframe above)
#install_github("seankross/warppipe")
library(warppipe)
OTU_ids <- seq_tbl("~/Dropbox/FISHPOO/unzipped/AllPoops/asvsforLULU_fisces_Nov2019.fasta")
head(OTU_ids)

#merging the OTU dataframe with the big main dataframe to get the sequences attached
test2 <- merge(alldat.mod, OTU_ids, by.x = 'Sq.id', by.y='Description')
#now make the Sequence the first column
new_df <- test2 %>%
  select(Sequence, everything())
#now make it back into CSV to delete the "Sq" column and the row #s column
write.csv(new_df,file="~/Dropbox/FISHPOO/unzipped/AllPoops/alldat_fisces_merge_Nov19LULU.csv",row.names=TRUE,quote=FALSE)

#read it back in
alldat_finalmod <- read.csv("~/Dropbox/FISHPOO/unzipped/AllPoops/alldat_fisces_merge_Nov19LULU_mod.csv")


#transform it for Assign Taxonomy function
alldat_final<-cbind(data.frame(t(alldat_finalmod)))
#Write CSV to make some more manual modifications (delete the X#s row at the top)
write.csv(alldat_final,file="~/Dropbox/FISHPOO/unzipped/AllPoops/alldat_fisces_finalmod_Nov19LULU.csv",row.names=TRUE,quote=FALSE)
#reading that modified file back in
alldat_FINAL <- read.csv("~/Dropbox/FISHPOO/unzipped/AllPoops/alldat_fisces_finalmod_Nov19LULU.csv", sep=",", row.names=1)

#Make it into a matrix so its formatted correctly for assigntaxonomy
LULUotu_fisces <- as.matrix(alldat_FINAL)
head(LULUotu_fisces)

#okay
```
Assign Taxonomy using LULU output (From DADA2 script from Carly)
```{r}

#DADA2 provides a native implementation of the RDP's naive Bayesian classifier. The assignTaxonomy function takes a set of sequences and a training set of taxonomically classified sequences, and outputs the taxonomic assignments with at least minBoot bootstrap confidence.
#Here, I am using a database from Ross Cunning (Cunning et al. 2017) modified for phyloseq (by adding >Symbiodinium; Clade XX; XX.1)
#
library(dada2); packageVersion("dada2"); citation("dada2")
library(ShortRead); packageVersion("ShortRead")
library(ggplot2); packageVersion("ggplot2")
library(phyloseq); packageVersion("phyloseq")
help(package="dada2")

path <- "~/Dropbox/FISHPOO/unzipped/AllPoops/"

#assigning taxononmy to the LULU data using DADA2 "assigntaxonomy" 
taxaLULU_fices <- assignTaxonomy(LULUotu_fisces,"~/Dropbox/FISHPOO/unzipped/AllPoops/ITS2db_trimmed_derep_LHK_phyloseqmodFINAL.fasta", minBoot=50,multithread=TRUE,tryRC=TRUE,outputBootstraps=FALSE)


unname(head(taxaLULU_fices, 30))
unname(taxaLULU_fices)


################################
##### output 'OTU' table #######
################################

#seqtab.nochim is the 'OTU' table...making this so that we can then assign taxonomy using blast, then modify the DADA2 taxonomy results from above to match the blast results


#First, output fasta file for 'OTUs'
path='~/Dropbox/FISHPOO/unzipped/AllPoops/LULUotu_fisces_Nov19.fasta'
uniquesToFasta(LULUotu_fisces, path, ids = NULL, mode = "w", width = 20000)

#making a new LULUotu with "ids" so that I have a copy with the sequences as headers (LULUotu_2) and a copy with short sq names as headers (LULUotu_2ids)
LULUotu_fisces_ids <- LULUotu_fisces


#then, rename output table and write it out
ids <- paste0("sq", seq(1, length(colnames(LULUotu_fisces))))
colnames(LULUotu_fisces)<-ids

head(LULUotu_fisces)
head(LULUotu_fisces_ids)

write.csv(LULUotu_fisces,file="LULUotu_fices_Nov19.csv",quote=F)

str(LULUotu_fisces_ids)

#Now, in terminal, run blast against Symbio database to compare
#makeblastdb -in ITS2db_trimmed_derep_LHK_phyloseqmodFINAL.fasta -dbtype nucl
#blastn -query LULUotu_fisces_Nov19.fasta -db ITS2db_trimmed_derep_LHK_phyloseqmodFINAL.fasta -num_alignments 1 -out BlastTable.LULU_fisces.txt -outfmt "6 qseqid sseqid pident nident evalue bitscore"

#put the blast output into excel, used vlookup to move blast result to replace DADA2 "assigntaxonomy" result

#################################
#Now, save outputs from DADA2 taxonomy so can come back to the analysis stage at a later point if desired
saveRDS(taxaLULU_fices, file="/Users/laurenhowe-kerr/Dropbox/FISHPOO/unzipped/AllPoops/taxaLULU_fices_Nov19.rds") ##making a new working assigntaxonomy output that I will next manually modify to meet Blast output

write.csv(taxaLULU_fices, file="/Users/laurenhowe-kerr/Dropbox/FISHPOO/unzipped/AllPoops/taxaLULU_fices_Nov19.csv") ##made this csv file with the same info as rds file, then changed csv in excel, now goign to save back as rds file)
LULU_fisces_blastmod <- as.matrix(read.csv("/Users/laurenhowe-kerr/Dropbox/FISHPOO/unzipped/AllPoops/taxaLULU_fices_Nov19_BLASTmod.csv", row.names = 1))
head(LULU_fisces_blastmod)

saveRDS(LULU_fisces_blastmod, file="/Users/laurenhowe-kerr/Dropbox/FISHPOO/unzipped/AllPoops/taxa_fisces_blastmod_Nov19LULU.rds")


#If you need to read in previously saved datafiles
LULU_fisces_blastmod <- readRDS("/Users/laurenhowe-kerr/Dropbox/FISHPOO/unzipped/AllPoops/taxa_fisces_blastmod_Nov19LULU.rds")
head(LULU_fisces_blastmod)


taxa.print.LULU_MESS <- LULU_fisces_blastmod # Removing sequence rownames for display only
rownames(taxa.print.LULU_MESS) <- NULL
head(taxa.print.LULU_MESS)


################################
##### handoff 2 phyloseq #######
################################

#import dataframe holding sample information (make sure Sample Names are the same as those attached to the OTU/seqtab table)
samdf<-read.csv("~/Dropbox/FISHPOO/unzipped/AllPoops/FISHPOO_variables_phyloseq.csv")
head(samdf)
rownames(samdf) <- samdf$Sample.ID

head(LULUotu_fisces_ids)
head(LULU_fisces_blastmod)

# Construct phyloseq object (straightforward from dada2 outputs)
LULUps <- phyloseq(otu_table(LULUotu_fisces_ids, taxa_are_rows=FALSE), 
               sample_data(samdf), 
               tax_table(LULU_fisces_blastmod))

LULUps

saveRDS(LULUps, file="~/Dropbox/FISHPOO/unzipped/AllPoops/LULUps_fisces_Nov19.rds")
```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
